xsd_validate/xsd_validate_activity_test.go

```diff
@@ -0,0 +1,118 @@
+
+import (
+	"path"
+	"path/filepath"
+	"runtime"
+	"testing"
+
+	temporalsdk_activity "go.temporal.io/sdk/activity"
+	temporalsdk_testsuite "go.temporal.io/sdk/testsuite"
+	"gotest.tools/v3/assert"
+	tfs "gotest.tools/v3/fs"
+
+	xsdvalidate "github.com/artefactual-sdps/temporal-activities/xsd_validate"
+)
+
+func TestXSDValidateActivity(t *testing.T) {
+	// Determine paths to test fixtures.
+	_, file, _, ok := runtime.Caller(0)
+	assert.Equal(t, ok, true)
+	fixture_folder_path := filepath.Join(path.Dir(file), "fixtures")
+
+	XSDTestFilePath := filepath.Join(fixture_folder_path, "person.xsd")
+	validXMLFilePath := filepath.Join(fixture_folder_path, "person_valid.xml")
+	invalidXMLFilePath := filepath.Join(fixture_folder_path, "person_invalid.xml")
+
+	// Assemble expected failures for XML invalid for the test XSD schema.
+	invalidXMLFailures := invalidXMLFilePath + `:3: element age: Schemas validity error : Element 'age': This element is not expected. Expected is ( name ).
+` + invalidXMLFilePath + ` fails to validate
+`
+
+	// Create file with invalid XSD and assemble expected failures.
+	invalidXSDFilePath := tfs.NewFile(t, "invalid_xsd_file", tfs.WithContent("junk")).Path()
+
+	invalidXSDFailures := invalidXSDFilePath + `:1: parser error : Start tag expected, '<' not found
+junk
+^
+Schemas parser error : Failed to parse the XML resource '` + invalidXSDFilePath + `'.
+WXS schema ` + invalidXSDFilePath + ` failed to compile
+`
+
+	// Create non-existent file.
+	nonExistentFile := tfs.NewFile(t, "removed_file")
+	nonExistentFile.Remove()
+
+	type test struct {
+		name    string
+		params  xsdvalidate.XSDValidateActivityParams
+		want    xsdvalidate.XSDValidateActivityResult
+		wantErr string
+	}
+	for _, tt := range []test{
+		{
+			name: "Test XSD validate activity with a valid XML file and a valid XSD file",
+			params: xsdvalidate.XSDValidateActivityParams{
+				XMLFilePath: validXMLFilePath,
+				XSDFilePath: XSDTestFilePath,
+			},
+			want: xsdvalidate.XSDValidateActivityResult{},
+		},
+		{
+			name: "Test XSD validate activity with an invalid XML file and a valid XSD file " + XSDTestFilePath,
+			params: xsdvalidate.XSDValidateActivityParams{
+				XMLFilePath: invalidXMLFilePath,
+				XSDFilePath: XSDTestFilePath,
+			},
+			want: xsdvalidate.XSDValidateActivityResult{Failures: invalidXMLFailures},
+		},
+		{
+			name: "Test XSD validate activity with a valid XML file and an invalid XSD file",
+			params: xsdvalidate.XSDValidateActivityParams{
+				XMLFilePath: validXMLFilePath,
+				XSDFilePath: invalidXSDFilePath,
+			},
+			want: xsdvalidate.XSDValidateActivityResult{Failures: invalidXSDFailures},
+		},
+		{
+			name: "Test XSD validate activity with a non-existent XML file and a valid XSD file",
+			params: xsdvalidate.XSDValidateActivityParams{
+				XMLFilePath: nonExistentFile.Path(),
+				XSDFilePath: XSDTestFilePath,
+			},
+			want:    xsdvalidate.XSDValidateActivityResult{},
+			wantErr: "no such file or directory",
+		},
```

Comment:
Here's a change that we can make to your subtests so we don't need to have global state:

1. Update the test struct as follows, so we can define functions that init the state and return the parameters:

    ```diff
    -params  xsdvalidate.XSDValidateActivityParams
    +params  func() xsdvalidate.XSDValidateActivityParams
    ```

2. The subtest can now be defined as follows:

	```go
	{
		name: "Test XSD validate activity with a valid XML file and a non-existent XSD file",
		params: func() xsdvalidate.XSDValidateActivityParams {
			nonExistentFile := tfs.NewFile(t, "removed_file")
			nonExistentFile.Remove()
			return xsdvalidate.XSDValidateActivityParams{
				XMLFilePath: filepath.Join("fixtures", "person_valid.xml"),
				XSDFilePath: nonExistentFile.Path(),
			}
		},
		want:    xsdvalidate.XSDValidateActivityResult{},
		wantErr: "no such file or directory",
	}
	```

3. Execute the `params` function to retrieve the activity parameters:

    ```diff
    -future, err := env.ExecuteActivity(xsdvalidate.XSDValidateActivityName, tt.params)
    +future, err := env.ExecuteActivity(xsdvalidate.XSDValidateActivityName, tt.params())
    ```

Reply:
(Done.)

Reply:
I made "want" a function too as it needed access to the activity params.